lightning/src/ln/functional_test_utils.rs

```diff
@@ -2255,7 +2235,16 @@ pub fn pass_failed_payment_back_no_abandon<'a, 'b, 'c>(origin_node: &Node<'a, 'b
 					payment_id.unwrap()
 				},
 				_ => panic!("Unexpected event"),
-			});
+			};
+			if i == expected_paths.len() - 1 {
+				match events[1] {
+					Event::PaymentFailed { ref payment_hash, ref payment_id } => {
+						assert_eq!(*payment_hash, our_payment_hash, "unexpected second payment_hash");
+						assert_eq!(*payment_id, expected_payment_id);
+					}
+					_ => panic!("Unexpected second event"),
+				}
+			}
```

Comment:
Not sure I completely understand the changes to the testing behavior. Not necessarily here -- which may be justified -- but elsewhere where the number of events has increased because of the additional `Event::PaymentFailed` generated. Wouldn't it be less invasive if the `Event::PaymentFailed` case was only checked in a dedicated test? Then when testing other behavior, we could have retries remaining such that no new event are generated so those tests are more narrowly focused.

Reply:
A portion of the changed tests won't have retryable HTLCs, so the `PaymentFailed` will be generated no matter what. Most of the others that I'm seeing tend to use the test utils such as `route_payment`, but when I update that util to use `_with_retry` I get 55 test failures.. I think it might be tricky since having retryable failed HTLCs will often lead to the generation of retry msg events to be handled. That said, it would probably be good in the long term to move the test utils to use `_with_retry`